ceph/rbd/workflows/rbd.py

```diff
@@ -36,7 +37,179 @@ def create_snap_and_clone(rbd, snap_spec, clone_spec, **kw):
+                    "is_secondary",
+                    "snap_schedule_levels",
+                    "snap_schedule_intervals",
+                    "io_size",
+                ]
+            }
+        )
+        create_config.update({"cluster": cluster})
+        out, err = rbd.create(**create_config)
+        if out or err:
+            log.error(f"Image {image} creation failed with err {out} {err}")
+            if raise_exception:
+                raise Exception(f"Image {image} creation failed with err {out} {err}")
+            return 1
+    return 0
+
+
+def create_images(
+    config,
+    cluster,
+    rbd,
+    pool,
+    pool_config,
+    image_config,
+    is_ec_pool,
+    create_image_parallely,
+):
+    """
+    Method to create multiple images parallely/sequentially
+    """
+    if not create_image_parallely:
+        for image, image_config_val in image_config.items():
+            rc = create_single_image(
+                config,
+                cluster,
+                rbd,
+                pool,
+                pool_config,
+                image,
+                image_config_val,
+                is_ec_pool,
+            )
+            if rc:
+                return rc
+    else:
+        try:
+            with parallel() as p:
+                for image, image_config_val in image_config.items():
+                    p.spawn(
+                        create_single_image,
+                        config,
+                        cluster,
+                        rbd,
+                        pool,
+                        pool_config,
+                        image,
+                        image_config_val,
+                        is_ec_pool,
+                        raise_exception=True,
+                    )
+        except Exception:
+            return 1
+    return 0
+
+
+def create_single_pool(
+    config,
+    pool,
+    pool_config,
+    client,
+    cluster,
+    rbd,
+    ceph_version,
+    is_ec_pool,
+    is_secondary,
+    create_image_parallely,
+    raise_exception=False,
+):
+    """
+    Method to create a single pool
+    """
+    if create_pool(
+        pool=pool,
+        pg_num=pool_config.get("pg_num", 64),
+        pgp_num=pool_config.get("pgp_num", 64),
+        client=client,
+        cluster=cluster,
+    ):
+        log.error(f"Pool creation failed for pool {pool}")
+        if raise_exception:
+            raise Exception(f"Pool creation failed for pool {pool}")
+        return 1
+
+    if ceph_version >= 3:
+        pool_init_conf = {"pool-name": pool, "cluster": cluster}
+        rbd.pool.init(**pool_init_conf)
+
+    if is_ec_pool and create_ecpool(
+        pool=pool_config.get("data_pool"),
+        k_m=pool_config.get("ec-pool-k-m"),
+        profile=pool_config.get("ec_profile", "use_default"),
+        pg_num=pool_config.get("ec_pg_num", 32),
+        pgp_num=pool_config.get("ec_pgp_num", 32),
+        failure_domain=pool_config.get("failure_domain", ""),
```

Comment:
what is this parameter `failure_domain`?

Reply:
It's one of the parameters used in ec pool. Please check doc.